### DISCUSSION TITLE
IPv6 support for Ingress controller

### DISCUSSION THREAD

**Discussion by kunalekawde** [nginx/kubernetes-ingress#3821] [Q&A] [answered]
Hello Folks,
1. Does ingress controller support dual stack in a way that external IPv4 or IPv6 traffic is processed and forwarded to application ?
2. Based on 1 above, if it receives IPv6 traffic can it then originate IPv4 traffic towards application by specifying the IPv4 kubernetes service in kind:ingress/ingress-rule ?

Thanks

**Comment by brianehlert** [comment, marked as answer]
First of all, I will always suggest that folks consider the CRDs this project offers. As it gives you much better controls, more options, and will introduce you to a pattern that is very similar to the evolving Gateway API efforts of the K8s community.

Dual stack is fully supported.
This applies to both the NGINX listener configuration and the assumptions we have around the container image.  We assume that both stacks are present in the container image.

The means the following combinations are possible:
client side <-> backend
IPv4 <-> IPv4
IPv6 <-> IPv6
IPv4 <-> IPv6
IPv6 <-> IPv4

The pattern we see today is that IPv6 is becoming popular / common for the client connection in regions such as Asia where they might not also be operating IPv6 internal to the cluster (for the backend services).

We included an option to disable IPv6 due to discovering that there are systems out there that disable IPv6 through disabling it within the kernel of the cluster nodes which creates some downstream impacts within the container.

**Reply by kunalekawde** [reply]
@brianehlert Thanks for the information. It's awesome to know that dual stack is supported. For my use case, I wanted to avoid application pod stacks being upgraded to dual stack support while only having ingress controller expose IPv6. kubernetes cluster / nodes will be dual stack support. 
As I understand from combination presented above, my use case should be covered.

If I may extend this, on the egress path (traffic originated from application pod to external servers), as I understood from one of video is that this can be achieved using nginx sidecar. App pod --> nginx sidecar --> nginx ingress controller --> out. 
Could you confirm if this is the case?  Can we get traffic directly from app pod to ingress controller?

**Reply by brianehlert** [reply]
The free project NGINX Service Mesh can be used to steer traffic that originates at a backend service out through the ingress controller.
That said, any request that comes in through the ingress controller will have the response go back out through the ingress controller.  No steering is necessary in that case.

The egress use cases are a bit unique, and I always ask for clarification about what someone is trying to achieve with an egress use case.

**Reply by kunalekawde** [reply]
Thanks @brianehlert, 
Egress use case is to reach out to external server or peer to fetch the information required to process the incoming request or the application pod is itself a client and based on some trigger has to send out the request. It has to go via egress and not directly from pod as we can then apply mTLS primarily. Also, traffic originated from multiple pods will then go via this egress pod(ingress controller here).

For egress traffic is steering it via sidecar mandatory or can it directly be targeted to ingress controller ? 
For traffic initiated from ingress controller as egress, does it support HTTP/2 or it will be HTTP/1.1 only ?

**Reply by brianehlert** [reply]
Ingress controller still behaves as a reverse proxy here, so the HTTP depends on the external service.  It is essentially negotiating the forwarded request.
It is not enforced like an incoming request is.

steering can be perfomed by sidecar (proxy forwarding to ingress for a specific known upstream name) or through network policy.
For example; OpenShift has the concept of an 'egress gateway' and can forward all requests originating within the cluster out a specific path.

**Reply by kunalekawde** [reply]
> Ingress controller still behaves as a reverse proxy here, so the HTTP depends on the external service. It is essentially negotiating the forwarded request. It is not enforced like an incoming request is.
> 
Cool, I suppose you are referring to some k8 resource for external service for which I shall check doc but if otherwise if external server supports HTTP/2, the outgoing traffic will be HTTP/2 is what I understand.
Could you please point me to some links for same ?

> steering can be perfomed by sidecar (proxy forwarding to ingress for a specific known upstream name) or through network policy. For example; OpenShift has the concept of an 'egress gateway' and can forward all requests originating within the cluster out a specific path.

I see, so there is some mapping to upstream name. I shall explore network policy. 

If our external servers (for egress traffic) are learnt dynamically over period of time, is it possible to program the ingress controller at runtime to make it aware of this new external server? Is NGINX Plus mandatory for this ?

**Reply by brianehlert** [reply]
ExternalName services can be used with the CRDs (VirtualServer / TransportServer): https://github.com/nginxinc/kubernetes-ingress/tree/v3.1.0/examples/custom-resources/externalname-services

Or with Ingress: https://github.com/nginxinc/kubernetes-ingress/tree/v3.1.0/examples/ingress-resources/externalname-services

**Reply by kunalekawde** [reply]
Thanks, per above documentation, this feature is only available in NGINX Plus. Which means for egress traffic outside cluster, to support HTTP/2, it needs commercial NGINX Plus?

**Reply by brianehlert** [reply]
Yes.  ExternalName support requires the upstream DNS resolver feature which is currently available with NGINX Plus only.  That is correct.

Dual Stack is both free or paid.

**Reply by kunalekawde** [reply]
In case we dont need DNS resolver because we have external IP addresses, does the solution path still remains same ?

**Reply by brianehlert** [reply]
I am honestly not entirely sure.
If for the externalname you set an IP instead of a DNS...  The business logic might still be looking for a resolver to set up.
I am honestly not entirely sure.
The best I can do is encourage you to try and see what happens.

**Reply by kunalekawde** [reply]
If I can squeeze in another clarification:
If our external servers (for egress traffic) are learnt dynamically over period of time, is it possible to program the ingress controller at runtime to make it aware of this new external server? Per reading, NGINX supports graceful reload but NGINX Plus is via API without any restart. If true, can application pass this new config and trigger reload of NGINX ?
 
Thanks again for all clarifications @brianehlert, that helps.

**Reply by brianehlert** [reply]
The NGINX Ingress implementation has a controller layer and configuration management system under the hood.
While you could interact with the NGINX Plus API directly, the controller processes would overwrite anything you did on the next update.
Essentially you would be introducing a two source of truth problem into a system that is not designed to support that.

If you are using NGINX Plus, simply using a DNS Name and the ExternalName capability - NGINX uses a second resolver function to resolve the targets for this particular upstream group.  This second resolver is unique from the base OS DNS resolver and updates the upstream group in a way that remains compatible with the controller processes.

NGINX does a few things when reloading a configuration.  Changes to nginx.conf generally require a reload is triggered by the modifying system.  There are 'hard' and 'soft' ways of doing this.  The 'soft' way keeps existing sessions running on existing worker processes and starts up new worker processes with the new configuration.  The 'old' worker processes are allowed to close naturally.
If there are repeated reloads, eventually the oldest worker processes are force killed.  Because it gets to be too much to maintain efficiently or it is assumed that the session is so long lived there is no natural close.

The N+ API introduces a different method that does not reload at all.  No new worker processes.

This NGINX Ingress Controller takes advantage of all that behavior and also has additional optimizations to take into account the more dynamic nature of a K8s environment (continuously changing upstreams).
